Disappearing edges

If you suspect this could be a bug, follow the template.

- What version of Dgraph are you using? 1.0.3


- Have you tried reproducing the issue with latest release? yes


- What is the hardware spec (RAM, OS)? 8GB, LINUX


- Steps to reproduce the issue (command/config used to run Dgraph).
quads := fmt.Sprintf(<%s> * * ., node_id)
if parent_id != nil {
quads += fmt.Sprintf("\n<%s> <%s> <%s> .", parent_id, edge, node_id)
}
_, err := txn.Mutate(bgctx, &api.Mutation{
CommitNow: true,
DelNquads: byte(quads),
})

- Expected behaviour and actual result.
after that, parent_id, will not show any edge references to other nodes until you create a new one, which tells me that they are not deleted but only sort of hidden.

```
server_1  | Got schema: [predicate:"roles" value_type:UID directive:REVERSE ]
zero_1    | 2018/03/03 07:39:33 oracle.go:84: purging below ts:20136, len(o.commits):2, len(o.aborts):0
server_1  | 2018/03/02 00:12:22 mutation.go:187: Done schema update predicate:"roles" value_type:UID directive:REVERSE 
server_1  | 2018/03/02 00:13:11 groups.go:296: Asking if I can serve tablet for: _share_hash_
server_1  | Got schema: [predicate:"roles" value_type:UID ]
```

after that I couldn't use Ratel after that, not sure if it's related to the below
```
server_1  | 2018/03/03 07:47:22 groups.go:454: Unable to sync memberships. Error: rpc error: code = DeadlineExceeded desc = context deadline exceeded
server_1  | 2018/03/03 07:47:22 groups.go:670: Error in oracle delta stream. Error: rpc error: code = Unknown desc = Node is no longer leader.

```

--- a/dgraph/cmd/bulk/loader.go
+++ b/dgraph/cmd/bulk/loader.go
@@ -267,6 +267,7 @@ func (ld *loader) mapStage() {
 	for i := range ld.mappers {
 		ld.mappers[i] = nil
 	}
+	ld.xids.EvictAll()
 	x.Check(ld.xidDB.Close())
 	ld.xids = nil
 	runtime.GC()

--- a/query/mutation.go
+++ b/query/mutation.go
@@ -75,6 +75,12 @@ func expandEdges(ctx context.Context, m *intern.Mutations) ([]*intern.DirectedEd
 			edgeCopy.Attr = pred
 			edges = append(edges, &edgeCopy)
 
+			// We only want to delete the pred from <uid> + <_predicate_> posting list if this is
+			// a SP* deletion operation. Otherwise we just continue.
+			if edge.Op == intern.DirectedEdge_DEL && string(edge.Value) != x.Star {
+				continue
+			}
+
 			e := &intern.DirectedEdge{
 				Op:     edge.Op,
 				Entity: edge.GetEntity(),

--- a/systest/mutations_test.go
+++ b/systest/mutations_test.go
@@ -927,6 +927,7 @@ func DeleteWithExpandAll(t *testing.T, c *client.Dgraph) {
 	assigned, err := c.NewTxn().Mutate(ctx, &api.Mutation{
 		SetNquads: []byte(`
 			_:a <to> _:b .
+			_:b <name> "b" .
 			_:a <to> _:c .
 			_:a <to> _:d .
 		`),
@@ -935,29 +936,6 @@ func DeleteWithExpandAll(t *testing.T, c *client.Dgraph) {
 	require.NoError(t, err)
 	auid := assigned.Uids["a"]
 	buid := assigned.Uids["b"]
-
-	q := `{
-		  me(func: has(to)) {
-			to {
-				uid
-			}
-		  }
-		}`
-	resp, err := c.NewTxn().Query(ctx, q)
-	require.NoError(t, err)
-
-	type Person struct {
-		Uid string   `json:"uid"`
-		To  []Person `json:"to"`
-	}
-
-	type Root struct {
-		Me []Person `json:"me"`
-	}
-
-	var r Root
-	json.Unmarshal(resp.Json, &r)
-	require.Equal(t, 3, len(r.Me[0].To))
 
 	_, err = c.NewTxn().Mutate(ctx, &api.Mutation{
 		DelNquads: []byte(`
@@ -968,21 +946,31 @@ func DeleteWithExpandAll(t *testing.T, c *client.Dgraph) {
 	})
 	require.NoError(t, err)
 
-	resp, err = c.NewTxn().Query(ctx, q)
-	require.NoError(t, err)
-	json.Unmarshal(resp.Json, &r)
-	require.Equal(t, 2, len(r.Me[0].To))
-
-	expandAllQuery := `{
-		  me(func: has(to)) {
-			expand(_all_) {
-				uid
-			}
+	q := `query test($id: string) {
+		  me(func: uid($id)) {
+			_predicate_
 		  }
 		}`
-	resp, err = c.NewTxn().Query(ctx, expandAllQuery)
+
+	type Preds struct {
+		Predicates []string `json:"_predicate_"`
+	}
+
+	type Root struct {
+		Me []Preds `json:"me"`
+	}
+
+	var r Root
+	resp, err := c.NewTxn().QueryWithVars(ctx, q, map[string]string{"$id": auid})
 	require.NoError(t, err)
 	json.Unmarshal(resp.Json, &r)
-	require.Equal(t, 1, len(r.Me))
-	require.Equal(t, 2, len(r.Me[0].To))
+	// S P O deletion shouldn't delete "to" .
+	require.Equal(t, 1, len(r.Me[0].Predicates))
+	require.Equal(t, "to", r.Me[0].Predicates[0])
+
+	// b should not have any _predicate_.
+	resp, err = c.NewTxn().QueryWithVars(ctx, q, map[string]string{"$id": buid})
+	require.NoError(t, err)
+	json.Unmarshal(resp.Json, &r)
+	require.Equal(t, 0, len(r.Me))
 }

--- a/systest/mutations_test.go
+++ b/systest/mutations_test.go
@@ -49,6 +49,7 @@ func TestSystem(t *testing.T) {
 	t.Run("set after delete for list", wrap(SetAfterDeletionListType))
 	t.Run("empty strings with exact", wrap(EmptyNamesWithExact))
 	t.Run("empty strings with term", wrap(EmptyRoomsWithTermIndex))
+	t.Run("delete with expand all", wrap(DeleteWithExpandAll))
 }
 
 func ExpandAllLangTest(t *testing.T, c *client.Dgraph) {
@@ -919,4 +920,69 @@ func EmptyRoomsWithTermIndex(t *testing.T, c *client.Dgraph) {
 		}`)
 	require.NoError(t, err)
 	require.Equal(t, `{"offices":[{"count(office.room)":1}]}`, string(resp.GetJson()))
+}
+
+func DeleteWithExpandAll(t *testing.T, c *client.Dgraph) {
+	ctx := context.Background()
+	assigned, err := c.NewTxn().Mutate(ctx, &api.Mutation{
+		SetNquads: []byte(`
+			_:a <to> _:b .
+			_:a <to> _:c .
+			_:a <to> _:d .
+		`),
+		CommitNow: true,
+	})
+	require.NoError(t, err)
+	auid := assigned.Uids["a"]
+	buid := assigned.Uids["b"]
+
+	q := `{
+		  me(func: has(to)) {
+			to {
+				uid
+			}
+		  }
+		}`
+	resp, err := c.NewTxn().Query(ctx, q)
+	require.NoError(t, err)
+
+	type Person struct {
+		Uid string   `json:"uid"`
+		To  []Person `json:"to"`
+	}
+
+	type Root struct {
+		Me []Person `json:"me"`
+	}
+
+	var r Root
+	json.Unmarshal(resp.Json, &r)
+	require.Equal(t, 3, len(r.Me[0].To))
+
+	_, err = c.NewTxn().Mutate(ctx, &api.Mutation{
+		DelNquads: []byte(`
+			<` + auid + `> <to> <` + buid + `> .
+			<` + buid + `> * * .
+		`),
+		CommitNow: true,
+	})
+	require.NoError(t, err)
+
+	resp, err = c.NewTxn().Query(ctx, q)
+	require.NoError(t, err)
+	json.Unmarshal(resp.Json, &r)
+	require.Equal(t, 2, len(r.Me[0].To))
+
+	expandAllQuery := `{
+		  me(func: has(to)) {
+			expand(_all_) {
+				uid
+			}
+		  }
+		}`
+	resp, err = c.NewTxn().Query(ctx, expandAllQuery)
+	require.NoError(t, err)
+	json.Unmarshal(resp.Json, &r)
+	require.Equal(t, 1, len(r.Me))
+	require.Equal(t, 2, len(r.Me[0].To))
 }

